src/shogun/modelselection/parameter_tree.cpp

```diff
@@ -0,0 +1,473 @@
+		}
+	}
+	else
+	{
+		if (m_parent->has(param))
+		{
+			m_param_mapping.insert(std::make_pair(param, value));
+			return true;
+		}
+	}
+
+	return false;
+}
+
+ParameterNode *ParameterNode::get_current() {
+	auto tree = new ParameterNode(*m_parent);
+
+	std::string param;
+
+	auto get_lambda = [&tree, &param](auto val) {
+		tree->attach(param, val);
+	};
+
+	for (auto& param_pair : m_param_mapping)
+	{
+		param = param_pair.first;
+		sg_any_dispatch(param_pair.second, sg_vector_typemap, shogun::None{}, get_lambda);
+	}
+	return tree;
+}
+
+CSGObject *ParameterNode::to_object(const std::shared_ptr<ParameterNode>& tree) {
+}
+
+ModelSelectionTree::ModelSelectionTree(CSGObject& model)
+    : first(true), m_counter(0), m_node_complete(false)
+{
+	m_parent = std::shared_ptr<CSGObject>(model.clone());
+	for (auto const& param : model.get_params())
+	{
+		if (auto* value = m_parent->get(param.first, std::nothrow))
+		{
+			create_node(param.first, value);
+			SG_UNREF(value);
+		}
+	}
+	m_current_node = m_nodes.begin();
+}
+
+void ModelSelectionTree::reset()
+{
+	for (const auto& param_pair : m_param_mapping)
+		m_param_iter[param_pair.first] = m_param_begin[param_pair.first];
+	m_node_complete = true;
+	m_current_param = m_param_mapping.end()->first;
+	m_param_set.clear();
+	m_counter = 0;
+	first = true;
+}
+
+void ModelSelectionTree::create_node(const std::string& name, CSGObject* obj)
+{
+	m_nodes.insert(std::make_pair(
+	    name, std::make_shared<ModelSelectionTree>(ModelSelectionTree(*obj))));
+}
+
+ParameterNode* ModelSelectionTree::get_next()
+{
+	SG_PRINT("ModelSelectionTree::get_next() - %s\n", get_parent_name());
+	auto tree = new ParameterNode(*m_parent);
+
+	std::string param;
+
+	auto get_lambda = [&tree, &param, this](auto val) {
+
+		// decltype(val.begin()) is a "hacky" way of getting the iterator type
+		// without registering RandomIterator<T> in type_case.h but need to pass
+		// SGVector by value each time
+		auto beginning = any_cast<decltype(val.begin())>(m_param_begin[param]);
+		auto current = any_cast<decltype(val.begin())>(m_param_iter[param]);
+		auto end = any_cast<decltype(val.begin())>(m_param_end[param]);
+
+		if (std::next(current)==end && current == beginning)
+		{
+			++m_counter;
+			if (m_param_set.find(
+					std::next(m_param_mapping.find(m_current_param))->first) ==
+				m_param_set.end() ||
+				std::next(m_param_mapping.find(m_current_param)) ==
+				m_param_mapping.end())
+			{
+				m_current_param =
+						std::prev(m_param_mapping.find(m_current_param))->first;
+				SG_PRINT("updating m_current_param from %s::%s to %s::%s\n", get_parent_name(), param.c_str(), get_parent_name(), m_current_param.c_str())
+			}
+			tree->attach(
+					param, *any_cast<decltype(val.begin())>(m_param_begin[param]));
+			return;
+		}
+
+		tree->attach(
+				param, *any_cast<decltype(val.begin())>(m_param_iter[param]));
+		SG_PRINT("CURRENT: %s, PARAM: %s\n", m_current_param.c_str(), param.c_str())
```

Comment:
you can make those prints DEBUG and then leave them in actually, might be useful for future debugging?

Reply:
yup, I'll do that in the end, I just don't want to set to debug level because then it becomes too verbose